### PR TITLE
daemon/notify: sanity check state entries

### PR DESCRIPTION
**Pull request by lucab**
This performs some basic sanity checks against notify-state entries.
Notably, it checks lenght and content of `FDNAME` labels.

Closes: https://github.com/lucab/libsystemd-rs/issues/116

### CODE REVIEW

**Comment by lucab** [comment]
@janderholm could you please have a look at this new logic?

**Comment by lucab** [comment]
@swsnr this one is now ready for review.

File: src/daemon.rs [hunk 1]
```diff
@@ -197,3 +199,34 @@ impl fmt::Display for NotifyState {
...
+    Ok(())
+}
+
+/// Validate an `FDNAME` according to systemd rules.
+///
+/// The name may consist of arbitrary ASCII characters except control
+/// characters or ":". It may not be longer than 255 characters.
+fn validate_fdname(fdname: &str) -> Result<(), SdError> {
+    if fdname.len() > 255 {
+        return Err(format!("overlong fdname '{}'", fdname).into());
```
Comment (swsnr) [inline src/daemon.rs:222]: Perhaps `fdname '{}' longer than 255 characters`?  This error message would be immediately actionable to a developer, whereas the current message requires at least one step through the documentation to find out what the maximum is.

File: src/daemon.rs [hunk 2]
```diff
@@ -197,3 +199,34 @@ impl fmt::Display for NotifyState {
...
 }
+
+/// Perform some basic sanity checks against state entries.
+fn sanity_check_state_entries(state: &[NotifyState]) -> Result<(), SdError> {
+    for (index, entry) in state.iter().enumerate() {
+        match entry {
+            NotifyState::Fdname(ref name) => validate_fdname(name),
+            _ => Ok(()),
+        }
+        .map_err(|e| format!("invalid notify state entry #{}: {}", index, e))?;
```
Comment (swsnr) [inline src/daemon.rs:210]: Can't we add a debug representation of the entry itself instead of its index?

**Review by swsnr — COMMENTED** [review]
I think the error message could be slightly improved, but other than that looks good.

File: src/daemon.rs (hunk 1 above)
Comment (lucab) [reply src/daemon.rs]: Good point, it read better that way.

File: src/daemon.rs (hunk 2 above)
Comment (lucab) [reply src/daemon.rs:210]: Mmh, I'm not a fan of adding a Debug entry in the middle of an error message; it is up to lower levels to add more details to pinpoint what's wrong. Currently a full error would look like this:
```
invalid notify state entry #0: invalid character ':' in fdname 'foo:bar'
```

If the index here does not carry useful information, we can even drop it. The nested error should already provide enough details anyway to identify the problem.

Comment (lucab) [reply src/daemon.rs:210]: Yes, that's the general idea: intermediate layers add some context about the flowpath, while the leaf error provide details about the fault.

**Pull request merged**